### DISCUSSION TITLE
Question about sensor data used in attitude controller

### DISCUSSION THREAD

**Discussion by minisol0175** [bitcraze/discussions#504] [Q&A] [answered]
Hello, 
I am a student doing research with Crazyflie.

The attitude controller uses gyro from sensorData_t.

I want to know if this gyro is the value coming directly from the sensor or is it filtered through something comp or EKF...

And I want to know what the sensfusion6UpdateQ function does.

Thank you!

**Comment by hmllr** [comment, marked as answer]
The gyro data is preprocesed slightly, first by subtracting a bias, then by aligning to the airframe and the gravity and passing through a low-pass filter. You find the code doing this here: [https://github.com/bitcraze/crazyflie-firmware/blob/master/src/hal/src/sensors_bmi088_bmp388.c#L320](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/hal/src/sensors_bmi088_bmp388.c#L320 )Afterwards it is used in the EKF.

The sensfusion6UpdateQ function is implementing an IMU and AHRS sensor fusion algorithm, it origins from [http://www.x-io.co.uk/open-source-ahrs-with-x-imu](http://www.x-io.co.uk/open-source-ahrs-with-x-imu)

Hope this helps you,
Hanna

**Reply by minisol0175** [reply]
Thank you!
It has been very helpful!
Let me ask you one more question...

In aligning to the airframe part, it have 3 IMU alignment Euler angles (IMU_PHI, IMU_THETA, IMU_PSI)
I can't find where these params are coming from.
May I know where you're coming from? And can you guess what it means?

**Reply by hmllr** [reply]
this function is actually not doing anything on the CF2, as the IMU is aligned (all angles are set to 0 here: https://github.com/bitcraze/crazyflie-firmware/blob/master/src/platform/interface/platform_defaults.h#L47). These parameters were introduced to be able to have an IMU with a different orientation (as on the flapper).
